src/psyclone/dynamo0p3.py

```diff
@@ -8222,25 +8414,30 @@ def diff_basis(self, function_space):
 
         :raises GenerationError: if the evaluator shape is not recognised.
         '''
-        if self._kern.eval_shape in VALID_QUADRATURE_SHAPES:
-            # We need differential basis functions for quadrature
-            diff_basis_name = get_fs_diff_basis_name(function_space)
-            self._arglist.append(diff_basis_name)
-
-        elif self._kern.eval_shape in VALID_EVALUATOR_SHAPES:
-            # We need differential basis functions for an evaluator,
-            # potentially for multiple target spaces. _kern.eval_targets is
-            # a dict where the values are 2-tuples of
-            # (FunctionSpace, argument).
-            for _, target in self._kern.eval_targets.items():
-                diff_basis_name = get_fs_diff_basis_name(function_space,
-                                                         on_space=target[0])
+        for shape in self._kern.eval_shapes:
+            if shape in VALID_QUADRATURE_SHAPES:
+                # We need differential basis functions for quadrature. A
+                # kernel stub won't have a name for the corresponding
+                # quadrature argument so we create one by appending the
+                # last part of the shape name to "qr_".
+                diff_basis_name = get_fs_diff_basis_name(
+                    function_space, qr_var="qr_"+shape.split("_")[-1])
                 self._arglist.append(diff_basis_name)
-        else:
-            raise GenerationError(
-                "Internal error: unrecognised evaluator shape ({0}). Expected "
-                "one of: {1}".format(self._kern.eval_shape,
-                                     VALID_EVALUATOR_SHAPES))
+
+            elif shape in VALID_EVALUATOR_SHAPES:
+                # We need differential basis functions for an evaluator,
+                # potentially for multiple target spaces. _kern.eval_targets is
+                # a dict where the values are 2-tuples of
+                # (FunctionSpace, argument).
+                for _, target in self._kern.eval_targets.items():
+                    diff_basis_name = get_fs_diff_basis_name(
+                        function_space, on_space=target[0])
+                    self._arglist.append(diff_basis_name)
+            else:
+                raise GenerationError(
```

Comment:
Oh yes. That pre-dates our introduction of the `InternalError` class. I've brought it up to date and added quotes around {0} but not {1} since that is a list.